### DISCUSSION TITLE
Implementation of QuTools QuTAG as a fast counter

### DISCUSSION THREAD

**Discussion by barkonaut** [Ulm-IQO/qudi-core#46] [Q&A]
Hi all,

I wanted to know if someone has implemented (or trying to ) the TimeTagger from QuTools GmbH as a fast counter hardware module. I am interested in using this in the Pulsed measurement toolchain.

Cheers!

**Comment by LukePiewalker** [comment]
Dear @barkonaut,

We did use and try to implement the now discontinued quTau a while ago. However, at some point we did not further pursue this, since for our measurements we wanted to have some "start-multistop" feature and I think there the quTau was severly limited in histogram length for this type of measurement. And "start-single stop" (where it can have longer histograms) was not really an option for the type of measurements.

Speaking for the people behind "qudi-iqo-modules" I think we have to dissapoint you in this regard. But we're happy to add it to the repo in case you can implement it.

Maybe someone else is out there who already did it for the quTag?

Kind regards
Lukas

**Reply by barkonaut** [reply]
Thanks @LukePiewalker for your reply. I have been trying to implement the device, but I was running into issues that I have narrowed down in the last days. 

When I am changing the value of the parameter "record length" in the Analysis tab of the Pulsed Measurement GUI, I run into problems and the TT is only giving Zeros as the output. Is this something that you have observed before with other implementations? Secondly, would you be willing to share the hardware file for the QuTau if it is still around somewhere? This is the first time that I am attempting to write a hardware module and there might be something very obvious that I have missed. 

Thanks!

**Comment by LukePiewalker** [comment]
Dear @barkonaut,

Sorry for the late reply. I was actually not doing the implementation, so I cannot really tell what kind of issues had been encountered on the way.

Regarding the file I am not to sure if it would help because of the following:
 From a quick glance at the docu of the quTag they seem to now provide some python wrapper already.
Do you use this wrapper or do you wrap the C functions yourself? Cause for the quTau this wrapper had not been there yet.

**Reply by barkonaut** [reply]
Hi @LukePiewalker ,

I am using the Python wrapper that the company provides. This works fine in a script, but I have not been able to make it work reliably in Qudi. 

Thanks, once again, for your input and comments!

**Comment by Neverhorst** [comment]
@barkonaut , did you manage to get it working by any chance?

**Reply by barkonaut** [reply]
Hi @Neverhorst, we have an implementation that works with scripts and Jupyter notebooks. We haven't tested this with the pulsed measurement toolchain yet. However, we are working on making it publicly available.

Sorry for the significantly late response!

**Comment by JBeyer-TUBAF** [comment]
May I ask about any progress and the current status of this again, @barkonaut ? We would like to include a QuTAG in our setup soon and any (also partial) code would surely help a lot!

**Reply by barkonaut** [reply]
@JBeyer-TUBAF , we have a hardware file that can be used with Qudi. I don't mind sharing it with you privately if you write to me. We are working on making it publicly available.
